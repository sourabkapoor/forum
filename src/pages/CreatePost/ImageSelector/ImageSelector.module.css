
.image-selector__file-image {
  max-width: 400px;
  max-height: 400px;
  display: block;
  margin-bottom: 16px;
  margin-top: 16px;
  margin: 0 auto;
  width: 100%;
}
.image-selector-file-selector {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.image-selector__button {
  background-color: #dfe1e3;
  font-size: 14px;
  font-weight: 700;
  letter-spacing: unset;
  line-height: 17px;
  text-transform: unset;
  min-height: 32px;
  min-width: 32px;
  padding: 4px 16px;
  /* cursor: not-allowed; */
  cursor: pointer;
  /* color: rgba(26, 26, 27, 0.5); */
  border-radius: 9999px;
  border: none;
}

.image-selector__buttons {
    display: flex;
    gap: 16px;
    justify-content: center;
    margin-top: 16px;
}

.image-selector__upload{
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 80px;
  border: 1px dashed var(--secondary-color);
  border-radius: 4px;
}