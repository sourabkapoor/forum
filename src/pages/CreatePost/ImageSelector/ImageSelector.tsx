import React, { useRef } from "react";
import CSSModules from "react-css-modules";
import styles from "./ImageSelector.module.css";

interface Props {
  selectedFile?: string;
  onSelectImage: (event: React.ChangeEvent<HTMLInputElement>) => void;
  handleTabChange: (value: string) => void;
  setSelectedFile: (value: string) => void;
}

const ImageSelector: React.FC<Props> = ({
  selectedFile,
  handleTabChange,
  setSelectedFile,
  onSelectImage,
}) => {

  function uploadImage() {
    selectedFileRef.current?.click()
  }
  const selectedFileRef = useRef<HTMLInputElement | null>(null);
  return (
    <div styleName="image-selector__upload">
      {selectedFile ? (
        <div styleName="image-selector__file-selector">
          <img styleName="image-selector__file-image" src={selectedFile} alt="user generated"/>
          <div styleName="image-selector__buttons">
            <button
              styleName="image-selector__button"
              onClick={() => handleTabChange("post")}
              data-testid="back-to-post-button"
            >
              Back To Post
            </button>
            <button
              styleName="image-selector__button"
              onClick={() => setSelectedFile("")}
              data-testid="remove-button"
            >
              Remove
            </button>
          </div>
        </div>
      ) : (
        <div styleName="image-selector__buttons">
          <button
            type="button"
            styleName="image-selector__button"
            onClick={uploadImage}
          >
            Upload
          </button>
          <input
            type="file"
            ref={selectedFileRef}
            hidden
            data-testid="file-input"
            onChange={onSelectImage}
          />
        </div>
      )}
    </div>
  );
};

export default CSSModules(ImageSelector, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "log",
});
