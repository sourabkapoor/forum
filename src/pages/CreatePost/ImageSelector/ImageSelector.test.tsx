import {
  fireEvent,
  queryByPlaceholderText,
  render,
  screen,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import each from "jest-each";
import ImageSelector from "./ImageSelector";
import React from "react";

describe("ImageSelector", () => {
  test("snapshot", () => {
    const mockProps = {
      selectedFile: "mockFile",
      handleTabChange: jest.fn(),
      setSelectedFile: jest.fn(),
      onSelectImage: jest.fn(),
    };
    const { asFragment } = render(<ImageSelector {...mockProps} />);

    expect(asFragment()).toMatchSnapshot();
  });
  test("back to post button", async () => {
    const mockProps = {
      selectedFile: "mockFile",
      handleTabChange: jest.fn(),
      setSelectedFile: jest.fn(),
      onSelectImage: jest.fn(),
    };

    render(<ImageSelector {...mockProps} />);
    const backToPostButton = screen.queryByTestId("back-to-post-button");

    await userEvent.click(backToPostButton as HTMLButtonElement);

    expect(mockProps.handleTabChange).toHaveBeenCalledTimes(1);
  });
  test("remove button", async () => {
    const mockProps = {
      selectedFile: "mockFile",
      handleTabChange: jest.fn(),
      setSelectedFile: jest.fn(),
      onSelectImage: jest.fn(),
    };

    render(<ImageSelector {...mockProps} />);
    const removeButton = screen.queryByTestId("remove-button");

    await userEvent.click(removeButton as HTMLButtonElement);

    expect(mockProps.setSelectedFile).toHaveBeenCalledTimes(1);
  });

  test("upload file", async () => {
    const mockOnSelectImage = jest.fn();
    const mockInputClick = jest.fn();

    const mockUseRef = jest.fn().mockReturnValue({
      current: {
        click: mockInputClick,
      },
    });
    const mockProps = {
      selectedFile: undefined,
      handleTabChange: jest.fn(),
      setSelectedFile: jest.fn(),
      onSelectImage: mockOnSelectImage,
    };

    jest.mock("react", () => {
      return {
        ...jest.requireActual("react"),
        useRef: mockUseRef,
      };
    });

    render(<ImageSelector {...mockProps} />);
    const fileInput = screen.queryByTestId("file-input");
    const uploadButton = screen.queryByText("Upload")

    const fakeFile = new File(["hello"], "hello.png", { type: "image/png" });

    await userEvent.upload(fileInput as HTMLInputElement, fakeFile);

    await userEvent.click(uploadButton as HTMLButtonElement);

    expect(mockOnSelectImage).toHaveBeenCalledTimes(1);
    expect(mockInputClick).toHaveBeenCalledTimes(1);
  });
});
