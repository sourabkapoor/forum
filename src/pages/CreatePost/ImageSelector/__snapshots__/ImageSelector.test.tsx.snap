// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSelector snapshot 1`] = `
<DocumentFragment>
  <div
    class="image-selector__upload"
  >
    <div
      class="image-selector__file-selector"
    >
      <img
        alt="user generated"
        class="image-selector__file-image"
        src="mockFile"
      />
      <div
        class="image-selector__buttons"
      >
        <button
          class="image-selector__button"
          data-testid="back-to-post-button"
        >
          Back To Post
        </button>
        <button
          class="image-selector__button"
          data-testid="remove-button"
        >
          Remove
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
